### PR TITLE
Dashboard

### PR DESCRIPTION
**Pull request by JeremieSamson**
URl to go to the dashboard is espace-adherent/tableau-de-bord
Still needs some CSS, preview below

dashboard

### CODE REVIEW

File: front/style/pages/_dashboard.scss [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        width: 199px;
+        height: 47px;
+        border: solid 1px #000000;
+        background-color: #ffffff;
+        padding: 10px;
+        margin: 10px 0;
+        text-transform: uppercase;
+    }
+
+    .dashboard_tag {
```
Comment (thomasbisignani) [inline front/style/pages/_dashboard.scss]: Can you use `-` instead of underscore here ?

File: front/style/pages/_dashboard.scss [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    .newblue {
+        color: $blue--dark;
+    }
+
+    .button {
+        width: 199px;
+        height: 47px;
+        border: solid 1px #000000;
+        background-color: #ffffff;
```
Comment (thomasbisignani) [inline front/style/pages/_dashboard.scss]: You should use the `$white` variable here

File: front/style/pages/_dashboard.scss [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    .newblue {
+        color: $blue--dark;
+    }
+
+    .button {
+        width: 199px;
+        height: 47px;
+        border: solid 1px #000000;
```
Comment (thomasbisignani) [inline front/style/pages/_dashboard.scss]: You should use the `$black` variable :
 `border: 1px solid $black;`

File: front/style/pages/_dashboard.scss [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        background-color: #ffffff;
+        padding: 10px;
+        margin: 10px 0;
+        text-transform: uppercase;
+    }
+
+    .dashboard_tag {
+        height: 24px;
+        border-radius: 12px;
+        background-color: #ebebeb;
```
Comment (thomasbisignani) [inline front/style/pages/_dashboard.scss]: You can use maybe the `$gray` here ?

You can check if similar chart colors are already defined here : https://github.com/EnMarche/en-marche.fr/blob/master/front/style/mixins/_colors.scss

File: front/style/pages/_dashboard.scss [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    div.list {
+        &-title {
+            height: 24px;
+            font-size: 20px;
+        }
+
+        &-subtitle {
+            height: 16px;
+            font-size: 14px;
+            color: #999999;
```
Comment (thomasbisignani) [inline front/style/pages/_dashboard.scss]: `color: $gray--muted;`

File: front/style/pages/_dashboard.scss [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        width: 800px;
+        margin-bottom: 50px;
+
+        .localization, .emails, .events, .interests {
+            margin-bottom: 30px;
+        }
+
+        .localization {
+            padding: 20px;
+            background-color: #eeeeee;
```
Comment (thomasbisignani) [inline front/style/pages/_dashboard.scss]: ` background-color: $gray--light;`

File: translations/messages.fr.yml [hunk 7]
```diff
@@ -73,6 +75,56 @@ adherent.change_email.expired_key: Votre demande de changement d'email est expir
...
+dashboard.show_members: Voir tous les adhérents
+dashboard.show_committees: Voir tous les comités
+dashboard.my_emails_sended: Les e-mails que j'ai envoyés
+dashboard.show_emails: Voir tous les emails
+dashboard.my_events: Les événements que j'ai créés
+dashboard.show_events: Voir tous les événements
+dashboard.my_internets: Les thématiques qui m'intéressent
+dashboard.shortcut.label: Raccourcis
+dashboard.shortcut.stats_space: Espace statistiques
+dashboard.shortcut.leaflets_posters: 🎨Tracts et posters
```
Comment (thomasbisignani) [inline translations/messages.fr.yml]: missing whitespace here no ? `🎨T` =>  `🎨 T`

File: translations/messages.fr.yml [hunk 8]
```diff
@@ -73,6 +75,56 @@ adherent.change_email.expired_key: Votre demande de changement d'email est expir
...
+dashboard.shortcut.my_documents: 📄Mes documents
```
Comment (thomasbisignani) [inline translations/messages.fr.yml]: Same here and L105, L106, L107

File: features/dashboard.feature [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+  Background:
+    Given the following fixtures are loaded:
+      | LoadAdherentData        |
+      | LoadTurnkeyProjectData  |
+      | LoadCitizenProjectData  |
+      | LoadEventData           |
+
+  # Adherent
+  Scenario: As an adherent, I can see information about me
```
Comment (thomasbisignani) [inline features/dashboard.feature]: IMO we can group scenarios by adherents / referents / animator / etc.. no ? 
We can gain for example the step call `Given I am logged as..`  and the `I am on...`

File: front/style/app.scss [hunk 10]
```diff
@@ -63,3 +63,4 @@
 @import "pages/join";
 @import "pages/complete";
 @import "pages/grande-marche-europe";
+@import "pages/dashboard";
```
[the remark this answers is not in the document]
Comment (thomasbisignani) [reply front/style/app.scss]: `adherent-dashboard` instead no ?

[the remark this answers is not in the document]
Comment (ottaviano) [reply front/style/app.scss]: yes 👍

File: src/Controller/EnMarche/AdherentController.php [hunk 11]
```diff
@@ -70,6 +74,31 @@ public function homeAction(Request $request, AdherentRepository $adherentReposit
...
+        EmailRepository $emailRepository,
+        SummaryRepository $summaryRepository,
+        MemberActivityTracker $memberActivityTracker
+    ): Response {
+        return $this->render('adherent/dashboard.html.twig', [
+            'events' => $eventRepository->findEventsByOrganizer($this->getUser()),
+            'emails' => $emailRepository->findBy(['sender' => $this->getUser()->getEmailAddress()]),
+            'summary' => $summaryRepository->findOneForAdherent($this->getUser()),
+            'activities' => $memberActivityTracker->getRecentActivitiesForAdherent($this->getUser()),
+            'area_stats' => $this->getUser()->getManagedArea()
```
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Controller/EnMarche/AdherentController.php]: @ottaviano does a referent always have a managed Area ?

File: src/Repository/AdherentRepository.php [hunk 12]
```diff
@@ -679,4 +680,54 @@ public function findAllInDistrict(District $district): Paginator
 
         return new Paginator($query, false);
     }
+
+    /**
+     * Finds enabled adherents in the deputy district.
+     *
+     * @return Adherent[]|Paginator
+     */
+    public function findAllInArea(District $district): Paginator
```
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Repository/AdherentRepository.php]: nice catch

File: features/dashboard.feature (hunk 9 above)
Comment (JeremieSamson) [reply features/dashboard.feature]: Yes and put the route in background also

File: src/Controller/EnMarche/AdherentController.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (ottaviano) [reply src/Controller/EnMarche/AdherentController.php]: yes, one referent must have a managed area filled

File: features/dashboard.feature [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+  # Referent
+  Scenario: As a referent, I can see information about me
+    Given I am logged as "referent@en-marche-dev.fr"
+    When I am on "/espace-adherent/tableau-de-bord"
+    Then I should see "Bienvenue, Referent !"
+    And I should see "adhérent depuis janvier 2017"
+    And I should see an "img" element
+
+    # As a referent,I can see information about my departement
```
Comment (thomasbisignani) [inline features/dashboard.feature]: missing whitespace here : `,I` => `, I`

File: src/Controller/EnMarche/AdherentController.php [hunk 14]
```diff
@@ -70,6 +74,31 @@ public function homeAction(Request $request, AdherentRepository $adherentReposit
...
+     */
+    public function dashboardAction(
+        AdherentRepository $adherentRepository,
+        EventRepository $eventRepository,
+        EmailRepository $emailRepository,
+        SummaryRepository $summaryRepository,
+        MemberActivityTracker $memberActivityTracker
+    ): Response {
+        return $this->render('adherent/dashboard.html.twig', [
+            'events' => $eventRepository->findEventsByOrganizer($this->getUser()),
```
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Controller/EnMarche/AdherentController.php]: done

File: src/Controller/EnMarche/AdherentController.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Controller/EnMarche/AdherentController.php]: done

File: src/Repository/AdherentRepository.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Repository/AdherentRepository.php]: removed

File: src/Repository/AdherentRepository.php [hunk 15]
```diff
@@ -679,4 +680,54 @@ public function findAllInDistrict(District $district): Paginator
...
+        ;
+
+        return new Paginator($query, false);
+    }
+
+    public function countInManagedArea(ReferentManagedArea $managedArea): int
+    {
+        return $this->createQueryBuilder('adherent')
+            ->select('COUNT(adherent)')
+            ->innerJoin('adherent.referentTags', 'tag')
```
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Repository/AdherentRepository.php]: it's not working like this, I have to use the join before

File: src/Repository/AdherentRepository.php [hunk 16]
```diff
@@ -679,4 +680,54 @@ public function findAllInDistrict(District $district): Paginator
...
+            ->getQuery()
+            ->getSingleScalarResult()
+        ;
+    }
+
+    public function countSubscriberInManagedArea(ReferentManagedArea $managedArea): int
+    {
+        return $this->createQueryBuilder('adherent')
+            ->select('COUNT(DISTINCT(adherent))')
+            ->innerJoin('adherent.referentTags', 'tag')
```
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Repository/AdherentRepository.php]: like the other, had to pass through innerJoin before

File: src/Repository/AdherentRepository.php [hunk 17]
```diff
@@ -679,4 +680,54 @@ public function findAllInDistrict(District $district): Paginator
...
+            ->innerJoin('adherent.subscriptionTypes', 'subscriptiontypes')
+            ->where('tag IN (:tags)')
+            ->andWhere('adherent.status = :status')
+            ->setParameter('tags', $managedArea->getTags())
+            ->setParameter('status', Adherent::ENABLED)
+            ->getQuery()
+            ->getSingleScalarResult()
+            ;
```
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Repository/AdherentRepository.php]: forgot this one, fixed in the last commit

File: front/style/pages/_dashboard.scss (hunk 1 above)
Comment (JeremieSamson) [reply front/style/pages/_dashboard.scss]: done

File: front/style/pages/_dashboard.scss (hunk 2 above)
Comment (JeremieSamson) [reply front/style/pages/_dashboard.scss]: done

File: front/style/pages/_dashboard.scss (hunk 3 above)
Comment (JeremieSamson) [reply front/style/pages/_dashboard.scss]: done

File: front/style/pages/_dashboard.scss (hunk 5 above)
Comment (JeremieSamson) [reply front/style/pages/_dashboard.scss]: done

File: front/style/pages/_dashboard.scss (hunk 4 above)
Comment (JeremieSamson) [reply front/style/pages/_dashboard.scss]: thanks, done

File: front/style/pages/_dashboard.scss (hunk 6 above)
Comment (JeremieSamson) [reply front/style/pages/_dashboard.scss]: done

File: translations/messages.fr.yml (hunk 7 above)
Comment (JeremieSamson) [reply translations/messages.fr.yml]: done

File: translations/messages.fr.yml (hunk 8 above)
Comment (JeremieSamson) [reply translations/messages.fr.yml]: done

File: features/dashboard.feature (hunk 9 above)
Comment (JeremieSamson) [reply features/dashboard.feature]: saw it together, done

File: features/dashboard.feature (hunk 13 above)
Comment (JeremieSamson) [reply features/dashboard.feature]: done

File: translations/messages.fr.yml [hunk 18]
```diff
@@ -73,6 +75,56 @@ adherent.change_email.expired_key: Votre demande de changement d'email est expir
...
+dashboard.show_events: Voir tous les événements
+dashboard.my_internets: Les thématiques qui m'intéressent
+dashboard.shortcut.label: Raccourcis
+dashboard.shortcut.stats_space: Espace statistiques
+dashboard.shortcut.leaflets_posters: 🎨 Tracts et posters
+dashboard.shortcut.my_documents: 📄 Mes documents
+dashboard.shortcut.contact: 🐛 Bug ? Nouveau besoin ? Faites-le nous savoir.
+dashboard.shortcut.procuration_manager_space: 🐼 Espace responsable procuration
+dashboard.shortcut.committee_coordinator_space: 🐸 Espace coordinateur de comité
+dashboard.shortcut.citizen_project_coordinator_space_space: 🐨Espace coordinateur de projet citoyen
```
[the remark this answers is not in the document]
Comment (JeremieSamson) [reply translations/messages.fr.yml]: done

File: front/style/pages/_adherent-dashboard.scss [hunk 19]
```diff
@@ -0,0 +1,97 @@
+.dashboard {
```
Comment (ottaviano) [inline front/style/pages/_adherent-dashboard.scss]: can you prefix with `adherent-` pls? `.dashboard` is too generic...

File: templates/adherent/dashboard.html.twig [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+{% block content %}
+    <section class="dashboard l__wrapper">
+        <div class="dashboard__welcome">
+            {% set imagePath = summary is not null ? path('asset_url', { path: summary.picturePath }) : '/images/user.png' %}
+
+            <p>
+            {% if summary is not null %}
+                <img src="{{ path('asset_url', { path: summary.picturePath }) }}" title="{{ 'dashboard.picture.update'|trans }}">
+            {% else %}
+                <a href="{{ path('app_summary_manager_step', {'step': 'photo'}) }}">
```
Comment (ottaviano) [inline templates/adherent/dashboard.html.twig]: pls can you remove `'` from the key? `{ step: 'photo' }`

File: templates/adherent/dashboard.html.twig [hunk 21]
```diff
@@ -0,0 +1,213 @@
+{% extends 'base.html.twig' %}
+
+{% block content %}
+    <section class="dashboard l__wrapper">
+        <div class="dashboard__welcome">
+            {% set imagePath = summary is not null ? path('asset_url', { path: summary.picturePath }) : '/images/user.png' %}
```
Comment (ottaviano) [inline templates/adherent/dashboard.html.twig]: where is `imagePath` used?

File: templates/adherent/dashboard.html.twig [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+                {% if app.user.isReferent or app.user.isDeputy or app.user.isSupervisor or app.user.isHost %}
+                    {{ get_user_level_label(app.user) }},
+                {% endif %}
+
+                {{ 'dashboard.adherent_since'|trans({'%date%': app.user.registeredAt|localizeddate('none', 'none', 'fr_FR', null, "MMMM Y")}) }}
+            </p>
+        </div>
+
+        <div class="dashboard__column-left">
+                {% if area_stats is not null %}
```
Comment (ottaviano) [inline templates/adherent/dashboard.html.twig]: CS: indent

File: templates/adherent/dashboard.html.twig [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+                            <li><a href="{{ path('app_referent_committees') }}" class="link--no-decor link--newblue">{{ 'dashboard.show_committees'|trans }}</a></li>
+                        </ul>
+                    </div>
+                {% endif %}
+                <div class="emails">
+                    <b>{{ 'dashboard.my_emails_sended'|trans }}</b>
+                    <hr>
+                    {% if emails|length > 0 %}
+                        <ul>
+                            {% for email in emails|slice(0,2) %}
```
Comment (ottaviano) [inline templates/adherent/dashboard.html.twig]: CS: `0,2` => `0, 2`

File: templates/adherent/dashboard.html.twig [hunk 24]
```diff
@@ -0,0 +1,213 @@
...
+                            {% endfor %}
+                        </ul>
+                    {% endif %}
+                </div>
+                <div class="events">
+                    <b>{{ 'dashboard.my_events'|trans }}</b>
+                    <hr>
+                    {% if events|length > 0 %}
+                        <ul>
+                            {% for event in events|slice(0,2) %}
```
Comment (ottaviano) [inline templates/adherent/dashboard.html.twig]: CS: `0, 2`

File: templates/adherent/dashboard.html.twig [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+                    {% if is_granted('ROLE_COORDINATOR_CITIZEN_PROJECT') %}
+                        <a href="{{ path('app_coordinator_citizen_project') }}" class="link--no-decor link--newblue">{{ 'dashboard.shortcut.citizen_project_coordinator_space_space'|trans }}</a>
+                    {% endif %}
+                    {% if app.user.isCitizenProjectAdministrator %}
+                        <li><a href="{{ path('app_citizen_project_image_generator') }}" class="link--no-decor link--newblue">{{ 'dashboard.shortcut.leaflets_posters'|trans }}</a></li>
+                    {% endif %}
+                    <li><a href="{{ path('app_documents_index') }}" class="link--no-decor link--newblue">{{ 'dashboard.shortcut.my_documents'|trans }}</a></li>
+
+                    {% if is_granted('ROLE_REFERENT') %}
+                        <li><a href="https://en-marche.helprace.com/" target="_blank" class="link--no-decor link--newblue">{{ 'dashboard.shortcut.contact'|trans }}</a></li>
```
Comment (ottaviano) [inline templates/adherent/dashboard.html.twig]: this help link should be displayed only for the referent users?

File: templates/adherent/dashboard.html.twig [hunk 26]
```diff
@@ -0,0 +1,213 @@
...
+                    {% endif %}
+                </ul>
+            </div>
+            <div class="skills">
+                {% set skillEditPath = path('app_summary_manager_step', ({step: 'skills'})) %}
+                <b>{{ 'dashboard.my_skills'|trans }}</b>
+                <hr>
+                {% if summary is not null and summary.skills|length > 0 %}
+                    <ul>
+                        {% for skill in summary.skills|slice(0,5) %}
```
Comment (ottaviano) [inline templates/adherent/dashboard.html.twig]: CS: `0, 5` and for other `slice` calls :)

File: templates/adherent/dashboard.html.twig (hunk 25 above)
Comment (JeremieSamson) [reply templates/adherent/dashboard.html.twig]: yes

File: front/style/pages/_adherent-dashboard.scss (hunk 19 above)
Comment (JeremieSamson) [reply front/style/pages/_adherent-dashboard.scss]: renamed

File: templates/adherent/dashboard.html.twig (hunk 20 above)
Comment (JeremieSamson) [reply templates/adherent/dashboard.html.twig]: done

File: templates/adherent/dashboard.html.twig (hunk 21 above)
Comment (JeremieSamson) [reply templates/adherent/dashboard.html.twig]: removed

File: templates/adherent/dashboard.html.twig (hunk 22 above)
Comment (JeremieSamson) [reply templates/adherent/dashboard.html.twig]: done

File: templates/adherent/dashboard.html.twig (hunk 23 above)
Comment (JeremieSamson) [reply templates/adherent/dashboard.html.twig]: done

File: templates/adherent/dashboard.html.twig (hunk 24 above)
Comment (JeremieSamson) [reply templates/adherent/dashboard.html.twig]: done

File: templates/adherent/dashboard.html.twig (hunk 26 above)
Comment (JeremieSamson) [reply templates/adherent/dashboard.html.twig]: done

File: features/adherent_dashboard.feature [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+    # As an adherent, I should have an interests section
+    Then I should see "Les thématiques qui m'intéressent"
+    And I should see "Aucun intérèt pour l'instant, renseignez-en ici"
+
+    # As an adherent, I should have 1 shortcut
+    Then I should see "Raccourcis"
+    And I should see 1 ".shortcuts ul li" elements
+    And I should see "Mes documents"
+
+    #As an adherent, I should have a skills section
```
Comment (thomasbisignani) [inline features/adherent_dashboard.feature]: missing whitespace here `#As` => `# As`

File: templates/adherent/dashboard.html.twig [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+            </div>
+
+            <div class="citizen-projects">
+                <b>{{ 'dashboard.my_citizen_projects'|trans }}</b>
+                <hr>
+                {% if app.user.citizenProjectMemberships|length > 0 %}
+                    <ul>
+                        {% for membership in app.user.citizenProjectMemberships|slice(0, 2) %}
+                            <li>
+                                {{  membership.citizenProject }}
```
Comment (thomasbisignani) [inline templates/adherent/dashboard.html.twig]: one too much whitespace after `{{`

**Pull request merged**